src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs

```diff
@@ -559,5 +559,37 @@ public override string ToString()
                 return this;
             }
         }
+
+        /// <summary>
+        /// Returns a new ServerDescription with a new HeartbeatException.
+        /// </summary>
+        /// <param name="heartbeatException">The heartbeat exception.</param>
+        /// <returns>
+        /// A new instance of ServerDescription.
+        /// </returns>
+        public ServerDescription WithHeartbeatException(Exception heartbeatException)
```

Comment:
This is a simplified version of the general purpose With method that is used while recovering from unexpected exceptions. The reason for having the simplified method is that it's possible the bug we are looking for is in the With method itself.